packages/core/src/builder.ts

```diff
@@ -0,0 +1,277 @@
+import * as token from "./token.js"
+import * as util from "./util.js"
+import Plugins from "./plugins.js"
+
+import { Fact, UcanPayload, isKeypair, Ucan, DidableKey, StoreI, BuilderI, BuildableState, CapabilityLookupCapableState, DefaultableState } from "./types.js"
+import { Capability, isCapability } from "./capability/index.js"
+import { capabilityCanBeDelegated, DelegationSemantics, DelegationChain } from "./attenuation.js"
+
+
+
+function isBuildableState(obj: unknown): obj is BuildableState {
+  return util.isRecord(obj)
+    && util.hasProp(obj, "issuer") && isKeypair(obj.issuer)
+    && util.hasProp(obj, "audience") && typeof obj.audience === "string"
+    && util.hasProp(obj, "expiration") && typeof obj.expiration === "number"
+}
+
+function isCapabilityLookupCapableState(obj: unknown): obj is CapabilityLookupCapableState {
+  return util.isRecord(obj)
+    && util.hasProp(obj, "issuer") && isKeypair(obj.issuer)
+    && util.hasProp(obj, "expiration") && typeof obj.expiration === "number"
+}
+
+type BuilderConstructor = new <State extends Partial<BuildableState>>(state: State, defaultable: DefaultableState) => BuilderI<State>
```

Comment:
Just a couple notes on this:
Looks like the function _has_ to be a default export for some reason: 
https://github.com/microsoft/TypeScript/issues/30355

& this type annotation needs to be included to suggest to the compiler how to write the declaration file

As well, I had to export some previously internal types in `./types` so the compiler was able to write the declaration file.

I was running into this error: https://github.com/microsoft/TypeScript/issues/5711